addons/web/static/src/public/public_component_service.js

```diff
@@ -0,0 +1,61 @@
+/** @odoo-module */
+
+import { App } from "@odoo/owl";
+import { registry } from "@web/core/registry";
+import { templates } from "@web/core/assets";
+import { _t } from "@web/core/l10n/translation";
+import { omit } from "@web/core/utils/objects";
+
+class ComponentManager {
+    constructor(env) {
+        this.env = env;
+        this.appConfig = {
+            templates,
+            env: env,
+            dev: env.debug,
+            translateFn: _t,
+            translatableAttributes: ["data-tooltip"],
+        };
+        /** @type {Map<HTMLElement, { app: App, mountProm: Promise<any> }>} */
+        this.apps = new Map();
+    }
+    mountComponents() {
+        for (const [key, component] of registry.category("public_components").getEntries()) {
+            for (const el of document.querySelectorAll(`[data-component="${CSS.escape(key)}"]`)) {
+                if (!this.apps.has(el)) {
+                    const props = {};
+                    if (el.dataset) {
+                        Object.assign(props, JSON.parse(el.dataset.props || "{}"));
+                        const dataset = omit(el.dataset, "component", "props");
+                        for (const [name, value] of Object.entries(dataset)) {
+                            props[name] = JSON.parse(value);
+                        }
+                    }
+                    const app = new App(component, {
+                        ...this.appConfig,
+                        props,
+                    });
+                    this.apps.set(el, { app, mountProm: app.mount(el) });
+                }
+            }
+        }
+        return Promise.all([...this.apps.values()].map(({ mountProm }) => mountProm)).then(() => {
+            // Don't expose the underlying apps
+            return;
+        });
```

Comment:
```suggestion
        await Promise.all([...this.apps.values()].map(({ mountProm }) => mountProm));
```

Reply:
That exposes the underlying apps.

Reply:
it's an `await`, not a `return`. The function will return `Promise<void>`

Reply:
Oh I missed that. Yeah this is better.